zipkin/src/main/java/com/linecorp/armeria/common/tracing/RequestContextCurrentTraceContext.java

```diff
@@ -0,0 +1,157 @@
+
+    public static final CurrentTraceContext INSTANCE = new RequestContextCurrentTraceContext();
+    private static final Logger logger = LoggerFactory.getLogger(RequestContextCurrentTraceContext.class);
+    private static final AttributeKey<TraceContext> TRACE_CONTEXT_KEY =
+            AttributeKey.newInstance("trace-context");
+
+    private static final Scope INCOMPLETE_CONFIGURATION_SCOPE = new Scope() {
+        @Override
+        public void close() {
+        }
+
+        @Override
+        public String toString() {
+            return "IncompleteConfigurationScope";
+        }
+    };
+
+    private static final Scope INITIAL_REQUEST_SCOPE = new Scope() {
+        @Override
+        public void close() {
+            // Don't remove the outer-most context (client or server request)
+        }
+
+        @Override
+        public String toString() {
+            return "InitialRequestScope";
+        }
+    };
+
+    /**
+     * Use this to ensure the trace context propagates to children.
+     *
+     * <p>Ex.
+     * <pre>{@code
+     *  // Ensure the trace context propagates to children
+     * ctx.onChild(RequestContextCurrentTraceContext::copy);
+     * }</pre>
+     */
+    public static void copy(RequestContext src, RequestContext dst) {
+        dst.attr(TRACE_CONTEXT_KEY).set(src.attr(TRACE_CONTEXT_KEY).get());
+    }
+
+    @Override
+    @Nullable
+    public TraceContext get() {
+        Attribute<TraceContext> traceContextAttribute = getTraceContextAttributeOrWarnOnce();
+        return traceContextAttribute != null ? traceContextAttribute.get() : null;
+    }
+
+    @Override
+    public Scope newScope(@Nullable TraceContext currentSpan) {
+        // Handle inspection added to ensure we can fail-fast if this isn't installed.
+        if (currentSpan != null && PingPongExtra.maybeSetPong(currentSpan)) {
+            return Scope.NOOP;
+        }
+
+        Attribute<TraceContext> traceContextAttribute = getTraceContextAttributeOrWarnOnce();
+        if (traceContextAttribute == null) {
+            return INCOMPLETE_CONFIGURATION_SCOPE;
+        }
+
+        final TraceContext previous = traceContextAttribute.getAndSet(currentSpan);
+
+        // Don't remove the outer-most context (client or server request)
+        if (previous == null) {
+            return INITIAL_REQUEST_SCOPE;
+        }
+
+        // Removes sub-spans (i.e. local spans) from the current context when Brave's scope does.
+        // If an asynchronous sub-span, it may still complete later.
+        class RequestContextTraceContextScope implements Scope {
+            @Override
+            public void close() {
+                // re-lookup the attribute to avoid holding a reference to the request if this scope is leaked
+                getTraceContextAttributeOrWarnOnce().set(previous);
+            }
+        }
+
+        return new RequestContextTraceContextScope();
```

Comment:
I think this can be simply replaced with a lambda `return () -> getTraceContextAttributeOrWarnOnce().set(previous);`. Do you have any reason for using local inner class?